src/Authentication.Abstractions/AzureSession.cs

```diff
@@ -263,16 +266,22 @@ public void UnregisterComponent<T>(string componentName) where T : class
             ChangeRegistry(
                 () =>
                 {
-                    var key = new ComponentKey(componentName, typeof(T));
-                    // todo: double check if dictionary operations are thread safe
-                    if (_componentRegistry.ContainsKey(key))
+                    sessionLock.EnterWriteLock();
+                    try
                     {
-                        var component = _componentRegistry[key];
-                        if (component is IAzureSessionListener listener)
+                        var key = new ComponentKey(componentName, typeof(T));
+                        if (_componentRegistry.TryGetValue(key, out var component))
                         {
-                            _eventHandler -= listener.OnEvent;
+                            if (component is IAzureSessionListener listener)
+                            {
+                                _eventHandler -= listener.OnEvent;
+                            }
+                            _componentRegistry.Remove(key);
```

Comment:
this line might also need atomic method applied

Reply:
Same as above